src/ModelContextProtocol.Core/Server/McpServerImpl.cs

```diff
@@ -365,23 +381,19 @@ await originalListResourceTemplatesHandler(request, cancellationToken).Configure
                 // Initial handler that sets MatchedPrimitive
                 if (request.Params?.Uri is { } uri && resources is not null)
                 {
-                    // First try an O(1) lookup by exact match.
+                    // Try an O(1) lookup by exact match.
                     if (resources.TryGetPrimitive(uri, out var resource))
                     {
                         request.MatchedPrimitive = resource;
                     }
                     else
                     {
-                        // Fall back to an O(N) lookup, trying to match against each URI template.
-                        // The number of templates is controlled by the server developer, and the number is expected to be
-                        // not terribly large. If that changes, this can be tweaked to enable a more efficient lookup.
+                        // Set MatchedPrimitive to the first templated resource as a hint.
+                        // The handler will validate if it actually matches.
                         foreach (var resourceTemplate in resources)
                         {
-                            // Check if this template would handle the request by testing if ReadAsync would succeed
                             if (resourceTemplate.IsTemplated)
                             {
-                                // This is a simplified check - a more robust implementation would match the URI pattern
-                                // For now, we'll let the actual handler attempt the match
```

Comment:
@halter73, the concept of MatchedPrimitive appears to be unsupportable with how we've defined resource template handling, where the actual primitive gets to decide if it can successfully handle the request. Are you imagining then that support would need to be moved out of the resource template and into this matching code?